anoblet/feat/jw-player

anoblet opened the pull request:
Known issues:

Reactivity - If the page is loaded with captions turned off, and then they are turned on via Storybook, the captions will fail to load. In a real world scenario this would not be an issue. (Fixed)

cu-task: https://app.clickup.com/t/apru1v

saas786 commented on packages/jw-player/src/mixins/ChapterMixin.js:
```diff
@@ -0,0 +1,127 @@
+
+export function ChapterMixin(BaseClass) {
+  class Mixin extends BaseClass {
+    __path = 'https://cxl.com/institute/wp-content/plugins/cxl-jwplayer/';
```
Hardcoded path, should be handled via PHP, using either:

`wp_add_inline_script` or `wp_localize_script`

anoblet replied:
I added the `plugin-path` attribute to specify this directory.

saas786 commented on packages/jw-player/src/mixins/ChapterMixin.js:
```diff
@@ -0,0 +1,127 @@
+    __addStyle() {
+      const style = document.createElement('style');
+      render(
+        css`
```
Inline css? can't be loaded from external file? Similar to how `cxl-ui` components load css?

anoblet replied:
I refactored the styles into a separate file.

saas786 commented on packages/jw-player/src/components/jw-player/index.css.js:
```diff
@@ -0,0 +1,89 @@
+  }
+
+  .captions span {
+    cursor: pointer;
```
```suggestion
    cursor: default;
```

Seeing hand / pointer on all page, makes it odd experience for me at least.

anoblet replied:
If an element performs an action, such as a button or a link, I would expect the cursor to turn into a pointer. This let's the user know that clicking the element will do something. I can get rid of the pointer, though it feels unnatural to me.

saas786 commented on packages/jw-player/src/plugins/telemetry-8.20.0.js:
```diff
@@ -0,0 +1,192 @@
+/* eslint-disable */
+
+// Rudder Stack Telemetry Plugin v8.1.0 Wed May 11 2022 02:20:30 GMT+1000 (Australian Eastern Standard Time)
```
Why we have this file?

Can't we use official sdk?
Ref: https://github.com/rudderlabs/rudder-sdk-js

Sample usage: https://github.com/rudderlabs/rudder-analytics-next/blob/4c3c6bb2d50c5c591c855d414832cb7c7fc1d390/sample-next-app-npm/pages/rudderInitialize.js#L2

anoblet replied:
This was submitted by one of our outside developers. We don't have much control right now what goes into the file.

saas786 replied:
> This was submitted by one of our outside developers. We don't have much control right now what goes into the file.
What you mean we don't have control? 
Its looks like a compiled / transpiled file, also no source of origin, where it was inspired from, based on which code would help too.

anoblet replied:
We just go access to the plugin source: https://github.com/conversionxl/cxl-jwplayer-plugins

In order for us to use/test it, it needs to be hosted:

https://github.com/conversionxl/aybolit/blob/1124d4b2f897320946e1dd26b1e0ba068bf411cf/packages/jw-player/src/components/jw-player/index.js#L21

anoblet commented:
@saas786 

>It case sensitive, should be case insensitive, so search for matching string in either case.

I don't believe the search is case sensitive. The default configuration options for `mark.js` have `caseSensitive` set to false. Would you mind letting me know which terms you are searching for?

>If finds multiple search matches, I think there should be "<" & ">" navigation to allow users to move between search matches.

Should we address this in a future PR?

>Should we have a playlist story? So we can see how it handles playlists?

Originally I had coded this for playlists though I was told that we would only be using media ids. I left the playlist functionality in there, and if you would like, I can create a story for it.

>Show chapters icon is misleading (to me at least), its a book mark icon.

I agree. CXL uses this icon currently. We could open a discussion to see if there is a better icon to use.

>Feature: I like how Youtube app recently started allowing Clips, so it will be nice to have a bookmark with start -> end positions, more like mini-clips from video, which user can refer to later on?

We might want to save this for a future PR.

>Otherwise looks good, found this error, can't see any log which shows the reason for 404 errors.

>Other error is there but not relevant I think.
See: https://developer.jwplayer.com/jwplayer/docs/jw8-player-errors-reference#334001

That looks like an intermittent connectivity issue.

saas786 commented:
@anoblet 

> > It case sensitive, should be case insensitive, so search for matching string in either case.
> I don't believe the search is case sensitive. The default configuration options for `mark.js` have `caseSensitive` set to false. Would you mind letting me know which terms you are searching for?

Word is `definitions`, I saw it as `Definitions`, it was not highlighted in init cap case.

> > If finds multiple search matches, I think there should be "<" & ">" navigation to allow users to move between search matches.
> Should we address this in a future PR?

Not critical, so can be in future PR, in case we decide to implement this.

> > Should we have a playlist story? So we can see how it handles playlists?
> Originally I had coded this for playlists though I was told that we wouldn't be using them. I left the playlist functionality in there, and if you would like, I can create a story for it.

As long as its possible, Kyle will have better grip on what we need, so we can skip playlist story I think.

> > Show chapters icon is misleading (to me at least), its a book mark icon.
> I agree. CXL uses this icon currently. We could open a discussion to see if there is a better icon to use.

Good, lets have a discussion on this.

> > Feature: I like how Youtube app recently started allowing Clips, so it will be nice to have a bookmark with start -> end positions, more like mini-clips from video, which user can refer to later on?
> We might want to save this for a future PR.

Sure, not a critical feature, but I find it useful, also can be a good UX, as some users does like to bookmark certain parts of videos, our customers might find it useful, if we can get it implemented.

> > Otherwise looks good, found this error, can't see any log which shows the reason for 404 errors.
> > Other error is there but not relevant I think.
> > See: [developer.jwplayer.com/jwplayer/docs/jw8-player-errors-reference#334001](https://developer.jwplayer.com/jwplayer/docs/jw8-player-errors-reference#334001)
> 
> That looks like an intermittent connectivity issue.

Yeah probably.

anoblet commented:
I fixed the caption reactivity which may have been the cause for the case-sensitivity while searching.

@saas786 Is there anything else that needs to be resolved before we merge this?

lkraav commented:
> I fixed the caption reactivity which may have been the cause for the case-sensitivity while searching.
> 
> @saas786 Is there anything else that needs to be resolved before we merge this?

I don't see any comments from @kylebrodeur?

saas786 commented:
> I fixed the caption reactivity which may have been the cause for the case-sensitivity while searching.
> 
> @saas786 Is there anything else that needs to be resolved before we merge this?
Will test it out again.

I will have another review / testing hopefully by tomorrow, so will update you afterwards.

saas786 commented:
> > It case sensitive, should be case insensitive, so search for matching string in either case.
> 
> I don't believe the search is case sensitive. The default configuration options for `mark.js` have `caseSensitive` set to false. Would you mind letting me know which terms you are searching for?

Should the heading also get highlighted?
![chrome_KOaX3OylYF](https://user-images.githubusercontent.com/1642796/176884202-f46965ad-8134-480c-bfd8-ec1ab584e5d8.png)

---

Tested playlist story: /?path=/story/jw-player-jw-player--playlist

Playlist icon looks odd to me, also says "more videos", which is confusing for me at least.
Did some search on playlist icons, but nothing handy found, so I guess icon is fine.

---

When I click on playlist, layout was odd for me:
![chrome_2tjxYKojwQ](https://user-images.githubusercontent.com/1642796/176886398-79fa7470-3f5e-4057-b22f-301b148df0c6.png)

Not sure its styles issue or what.

Should have be more like this:
![chrome_ICCOXmBV9N](https://user-images.githubusercontent.com/1642796/176886478-fbf34b0b-dd1e-45d7-816e-5bfa51a5332a.png)

saas786 commented:
@kyle 
I don't have access to: https://github.com/conversionxl/cxl-jwplayer-plugins

I want to see how code is organized, so whether we can publish it to https://npm.cxl.co/ or not.

As adding compiled / transpiled file directly to `aybolit` looks odd to me.

Ref: https://github.com/conversionxl/aybolit/blob/b3abc262006b630e440843bcdcc8fb3924afa0af/packages/jw-player/src/plugins/telemetry-8.20.0.js

Specifically when we need it hosted somewhere to be able to use in JW-Player.

Ref: https://github.com/conversionxl/aybolit/blob/b3abc262006b630e440843bcdcc8fb3924afa0af/packages/jw-player/src/components/jw-player/index.js#L21

So we can either npm i if needed to have it in compiled js files or maybe directly reference it from https://npm.cxl.co

kylebrodeur commented:
More videos screen shouldn't be showing up at all. There's no technical
need for it or use.

On Fri, Jul 1, 2022, 6:44 AM saasfreelancer ***@***.***>
wrote:

> @kyle <https://github.com/kyle>
> I don't have access to:
> https://github.com/conversionxl/cxl-jwplayer-plugins
>
> I want to see how code is organized, so whether we can publish it to
> https://npm.cxl.co/ or not.
>
> As adding compiled / transpiled file directly to aybolit looks odd to me.
>
> Ref:
> https://github.com/conversionxl/aybolit/blob/b3abc262006b630e440843bcdcc8fb3924afa0af/packages/jw-player/src/plugins/telemetry-8.20.0.js
>
> Specifically when we need it hosted somewhere to be able to use in
> JW-Player.
>
> Ref:
> https://github.com/conversionxl/aybolit/blob/b3abc262006b630e440843bcdcc8fb3924afa0af/packages/jw-player/src/components/jw-player/index.js#L21
>
> So we can either npm i if needed to have it in compiled js files or maybe
> directly reference it from https://npm.cxl.co
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/conversionxl/aybolit/pull/210#issuecomment-1172257690>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAGGAXYGY25WVCF5N62S7JLVR3KZ7ANCNFSM5XR6U42Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

kylebrodeur commented:
The plugins are a tad better in the mixin strategy than the file loading
strategy. There's a lot more hacks required to load it in PHP so having it
part of component is better, not just referencing files.

The plugins aren't that performance and they were made better under mix in
strategy.

Syed, I'll add you to the jwplaywr repo.

On Fri, Jul 1, 2022, 7:27 AM Kyle Brodeur ***@***.***> wrote:

> More videos screen shouldn't be showing up at all. There's no technical
> need for it or use.
>
> On Fri, Jul 1, 2022, 6:44 AM saasfreelancer ***@***.***>
> wrote:
>
>> @kyle <https://github.com/kyle>
>> I don't have access to:
>> https://github.com/conversionxl/cxl-jwplayer-plugins
>>
>> I want to see how code is organized, so whether we can publish it to
>> https://npm.cxl.co/ or not.
>>
>> As adding compiled / transpiled file directly to aybolit looks odd to me.
>>
>> Ref:
>> https://github.com/conversionxl/aybolit/blob/b3abc262006b630e440843bcdcc8fb3924afa0af/packages/jw-player/src/plugins/telemetry-8.20.0.js
>>
>> Specifically when we need it hosted somewhere to be able to use in
>> JW-Player.
>>
>> Ref:
>> https://github.com/conversionxl/aybolit/blob/b3abc262006b630e440843bcdcc8fb3924afa0af/packages/jw-player/src/components/jw-player/index.js#L21
>>
>> So we can either npm i if needed to have it in compiled js files or maybe
>> directly reference it from https://npm.cxl.co
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/conversionxl/aybolit/pull/210#issuecomment-1172257690>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AAGGAXYGY25WVCF5N62S7JLVR3KZ7ANCNFSM5XR6U42Q>
>> .
>> You are receiving this because you were mentioned.Message ID:
>> ***@***.***>
>>
>

kylebrodeur commented:
Search shouldn't be case sensitive. Should be able to find what you need in
any case.

On Fri, Jul 1, 2022, 7:28 AM Kyle Brodeur ***@***.***> wrote:

> The plugins are a tad better in the mixin strategy than the file loading
> strategy. There's a lot more hacks required to load it in PHP so having it
> part of component is better, not just referencing files.
>
> The plugins aren't that performance and they were made better under mix in
> strategy.
>
> Syed, I'll add you to the jwplaywr repo.
>
> On Fri, Jul 1, 2022, 7:27 AM Kyle Brodeur ***@***.***> wrote:
>
>> More videos screen shouldn't be showing up at all. There's no technical
>> need for it or use.
>>
>> On Fri, Jul 1, 2022, 6:44 AM saasfreelancer ***@***.***>
>> wrote:
>>
>>> @kyle <https://github.com/kyle>
>>> I don't have access to:
>>> https://github.com/conversionxl/cxl-jwplayer-plugins
>>>
>>> I want to see how code is organized, so whether we can publish it to
>>> https://npm.cxl.co/ or not.
>>>
>>> As adding compiled / transpiled file directly to aybolit looks odd to
>>> me.
>>>
>>> Ref:
>>> https://github.com/conversionxl/aybolit/blob/b3abc262006b630e440843bcdcc8fb3924afa0af/packages/jw-player/src/plugins/telemetry-8.20.0.js
>>>
>>> Specifically when we need it hosted somewhere to be able to use in
>>> JW-Player.
>>>
>>> Ref:
>>> https://github.com/conversionxl/aybolit/blob/b3abc262006b630e440843bcdcc8fb3924afa0af/packages/jw-player/src/components/jw-player/index.js#L21
>>>
>>> So we can either npm i if needed to have it in compiled js files or
>>> maybe directly reference it from https://npm.cxl.co
>>>
>>> —
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/conversionxl/aybolit/pull/210#issuecomment-1172257690>,
>>> or unsubscribe
>>> <https://github.com/notifications/unsubscribe-auth/AAGGAXYGY25WVCF5N62S7JLVR3KZ7ANCNFSM5XR6U42Q>
>>> .
>>> You are receiving this because you were mentioned.Message ID:
>>> ***@***.***>
>>>
>>

saas786 commented:
> The plugins are a tad better in the mixin strategy than the file loading
strategy. There's a lot more hacks required to load it in PHP so having it
part of component is better, not just referencing files.
> The plugins aren't that performance and they were made better under mix in
strategy.

We still need to import it via `npm i` instead of manually pushing it to `aybolit`. When `jw-player` component will be built, it can than import it into its compiled / transpiled code.

saas786 commented:
> Search shouldn't be case sensitive. Should be able to find what you need in any case.
It works like this, but it does exclude caption chapter title, so I was asking whether it should highlight the chapter title if it found relevant keyword or not.

anoblet commented:
@saas786 @kylebrodeur 

1) Updated the markjs selector to include chapter titles
2) Removed the telemetry plugin for now, until we can figure out where we want the code to reside and how we are going to host it.
3) Since playlists won't be used for the time being, I can either remove the story or we can just ignore it. It won't affect the default functionality.
4) Case sensitivity shouldn't be an issue anymore. Could someone please verify this?

saas786 commented:
> 4. Case sensitivity shouldn't be an issue anymore. Could someone please verify this?
Not an issue.

saas786 commented on packages/jw-player/src/lib/cxl-jwplayer-styles.js:
```diff
@@ -0,0 +1,195 @@
+import { css } from 'lit';
+
+export const cxlJWPlayerStyles = css`
```
@anoblet same here, maybe `scss` file should be used, and generate this css like `cxl-ui` does?

saas786 commented on packages/jw-player/src/mixins/chapter/index.css.js:
```diff
@@ -0,0 +1,59 @@
+import { css } from 'lit';
+
+export const style = css`
```
@anoblet same here, maybe create `scss`, generate css like how `cxl-ui` does it?

anoblet commented:
@saas786 I'm not sure using SCSS is necessary. The maximum depth of selectors is 2. There isn't much of an improvement with code clarity by switching over.

saas786 commented:
> @saas786 I'm not sure using SCSS is necessary. The maximum depth of selectors is 2. There isn't much of an improvement with code clarity by switching over.

This is not the primary reason, applying same code organization, and logic is the key reason. We can't be approaching same features differently in same codebase, unless there is some good reason.

anoblet commented:
@saas786 I switched to SCSS.

saas786 commented on packages/jw-player/src/mixins/BaseMixin.js:
```diff
@@ -0,0 +1,116 @@
+     */
+
+    // eslint-disable-next-line class-methods-use-this, no-unused-vars, no-empty-function
+    async __onTimeListener(event) {}
+
+    __registerListeners() {
+      this.__boundOnTimeListener = this.__onTimeListener.bind(this);
+      this.__jwPlayer.on('time', this.__boundOnTimeListener);
+    }
```
```suggestion
    async __onTimeListener(event) {}

    __registerListeners() {
      this.__jwPlayer.on('time', this.__onTimeListener);
    }
```

Can we also throttle it, as it seems like its being triggered 10 times per second.

Ref: https://developer.jwplayer.com/jwplayer/docs/jw8-javascript-api-reference#jwplayerontime

anoblet replied:
We need the binding here. This won't work:

// this.__boundOnTimeListener = this.__onTimeListener.bind(this);
this.__jwPlayer.on('time', this.__onTimeListener);

saas786 commented:
@anoblet 
Where are we doing a bundle / build for jwplayer?
after `yarn dist` I don't a bundle generated for it? Nor its style files are compiled to `css`.

In storybook its fine.

saas786 commented on packages/jw-player/package.json:
```diff
@@ -0,0 +1,17 @@
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
```
```suggestion
  "license": "SEE LICENSE IN LICENSE",
```

saas786 commented:
Did an apply formatting & linting on `jw-player` package?

It seems like it didn't get applied?

Ran these, and each either fixed, or warned about issues.

`npx prettier --single-quote=false --write "./packages/jw-player/scss"`
`npx stylelint --syntax=scss "./packages/jw-player/scss"`
`npx prettier --write "./packages/jw-player/src"`
`npx eslint --fix "./packages/jw-player/src"`
 
 Please apply these too.
 
 Also you please take a look at build, and see why they didn't get applied automatically on commit?

anoblet commented:
I moved `jw-player` into `cxl-ui` in order to get webpack mix to run correctly.

This file is legacy code and was ignored by the linting process: https://github.com/conversionxl/aybolit/blob/anoblet/feat/jw-player/packages/cxl-ui/scss/jw-player/legacy.scss

There's a fair amount of refactoring needing to be done on that file, and I'm not familiar enough with the expected result to do that myself.

Everything else has been linted correctly.

saas786 commented:
> I moved `jw-player` into `cxl-ui` in order to get webpack mix to run correctly.
@anoblet 
Should not have moved the top commit, should have set first commit. I see it can be an issue for you to rebase and such.
I will look into this ASAP, and see if I can clean it up, so by default `jw-player` is added to `cxl-ui` and later commits are applied on that.

Will also try to lint it myself. But that could be take a week, I am off next week, maybe you can take a digg on this yourself during this time.

saas786 commented:
> @saas786 The component is fairly decoupled. If you would like, I can clean this up and create only two commits:
> 
> 1. feat(cxl-ui): add jw-player component
> 2. feat(cxl-ui): update webpack.mix.js
> 
> I had to update the mix configuration to not throw on `stream` not being available.
> 
> [#210 (files)](https://github.com/conversionxl/aybolit/pull/210/files#diff-d2236e97fa3a42efd63b77322f8c1ea6ece4bb990f8b954b0853eb1e50e59b68)
> 
> I will also do my best to go through `legacy.scss` and see what is still needed. Reorganize it so that it passes the lint test.

I am squashing it into single commit, and will also do linting

saas786 commented:
> > @saas786 The component is fairly decoupled. If you would like, I can clean this up and create only two commits:
> > 
> > 1. feat(cxl-ui): add jw-player component
> > 2. feat(cxl-ui): update webpack.mix.js
> > 
> > I had to update the mix configuration to not throw on `stream` not being available.
> > [#210 (files)](https://github.com/conversionxl/aybolit/pull/210/files#diff-d2236e97fa3a42efd63b77322f8c1ea6ece4bb990f8b954b0853eb1e50e59b68)
> > I will also do my best to go through `legacy.scss` and see what is still needed. Reorganize it so that it passes the lint test.
> 
> I am squashing it into single commit, and will also do linting

It seems like you have already done the linting :)

Tested after squashing and seems to work fine.

saas786 commented:
Should we be showing search bar by default? I think there should be some toggle to show / hide it.

![MFWDcXhnuL](https://user-images.githubusercontent.com/1642796/179504997-77b531f1-c668-4af8-81e7-e80092a3b57d.png)

saas786 commented:
Apart from UX suggestions, I think it looks fine, just can't test it out on actual website using PHP. 
So no feedback on that.

lkraav commented:
> Should we be showing search bar by default? 

Yes, show by default.

anoblet commented:
I got rid off chapter navigation scroll bar styles by accident. Adding them back now.

![image](https://user-images.githubusercontent.com/7674171/179508629-979a45ce-e37f-4ff8-a906-84009ac26707.png)

lkraav commented:
> I got rid of chapter navigation scroll bar styles by accident. Adding them back now.

cxl-app-layout has on-brand scrollbar styles. Can you check if those are system-wide usable with cxl-lumo-styles, or maybe they should be migrated from cxl-app-layout for re-use?

anoblet commented:
Refactored to use:

```
@use "~@conversionxl/cxl-lumo-styles/scss/mixins";
...
@include mixins.better-webkit-scrollbars();
```

anoblet commented:
I mad some small changes to the search bar.

1) Switch to `vaadin-text-field`
2) Added search and clear icons
3) Hide matches until user has entered text and pressed enter
4) Switch to `vaadin-check-box` and use CXL red for the background color.

Do we want to listen to the input event as opposed to the change event? We could start highlighting matches after the user has entered a minimum number of characters (3?).

lkraav commented:
> Do we want to listen to the input event as opposed to the change event? We could start highlighting matches after the user has entered a minimum number of characters (3?).

Depends on performance, but yes, 3 char minimum sounds viable.

I've sometimes in forums found I need to search for 2 chars, but maybe it's not needed here for a large use case %?

saas786 commented:
> ![image](https://user-images.githubusercontent.com/7674171/179968128-85c186f2-5d4d-49fb-83b4-d6db4e8ed5fb.png)
> 
> I mad some small changes to the search bar.

Looks better to me atleast.

anoblet commented:
@lkraav @saas786 

Is there anything else we need to do to get this over the finish line?

lkraav commented:
> Is there anything else we need to do to get this over the finish line?

Does text search actually work? In Firefox, I'm getting

```
12:21:33.020 Uncaught TypeError: event.path is undefined
    __search https://deploy-preview-210--conversionxl-aybolit.netlify.app/main.a47ad19a.iframe.bundle.js:2
    invokeFunc https://deploy-preview-210--conversionxl-aybolit.netlify.app/vendors~main.9f1891a2.iframe.bundle.js:50
    leadingEdge https://deploy-preview-210--conversionxl-aybolit.netlify.app/vendors~main.9f1891a2.iframe.bundle.js:50
    debounced https://deploy-preview-210--conversionxl-aybolit.netlify.app/vendors~main.9f1891a2.iframe.bundle.js:50
    handleEvent https://deploy-preview-210--conversionxl-aybolit.netlify.app/vendors~main.9f1891a2.iframe.bundle.js:50
main.a47ad19a.iframe.bundle.js:2:469907
```

Story https://deploy-preview-210--conversionxl-aybolit.netlify.app/iframe.html?args=&id=jw-player-jw-player--default&viewMode=story

Everything else looks pretty good. @kylebrodeur can you take a look at Story today? This looks good and could move forward to PHP implementation.

lkraav commented:
Chapter navigation: is there opportunity for vaadin-dialog here?

![image](https://user-images.githubusercontent.com/147228/180409391-ca7b17ec-e5d4-43a4-a0c1-85491da5b686.png)

In current div, close button is in an unnatural location, small, and hard to see (low contrast). Having a standardized modal would make much more sense to my eye.

anoblet commented:
I fixed the search for Firefox. There's an issues when using `vaadin-text-field`, `@input`, and `debounce` where `event.target` would switch from the input to the custom element randomly. I used `event.path`, though did not test on Firefox. I switched to using `@query` which works in Chrome/Firefox (There's no Windows client for Safari).

I made some minor style updates to the chapter list, making the close button more prominent and in a more familiar position.

![image](https://user-images.githubusercontent.com/7674171/180459060-b2c8885f-58ff-486d-8f0b-319ea93df005.png)

I'm not sure I can visualize how a modal would work, could you please explain?

As for the more videos styling issue, I was told that we would not be using playlists for the time being (https://github.com/conversionxl/aybolit/pull/210#issuecomment-1172293454). I can fix the styling though if it turns out we need it.

lkraav commented:
> I'm not sure I can visualize how a modal would work, could you please explain?

vaadin-dialog in the middle of screen. Not sure there's anything else to imagine. But on phone rn, so can't really browse vaadin element demos.

saas786 commented:
> @lkraav
> 
> I fixed the search for Firefox. There's an issue when using `vaadin-text-field`, `@input`, and `debounce` where `event.target` would switch from the input to the custom element randomly. I used `event.path`, though did not test on Firefox. I switched to querying the element, and getting the value rather than relying on the event which I've tested in Chrome/Firefox (There's no Windows client for Safari) and it works.

Doesn't this work?
```
const path = e.composedPath  
    ? e.composedPath()  
    : (e.path || []);  
```

saas786 commented:
> I fixed the caption reactivity which may have been the cause for the case-sensitivity while searching.
> 
> @saas786 Is there anything else that needs to be resolved before we merge this?

Looks good. Can be further tweaked / improved (if needed later), for initial release looks good to me.

anoblet commented:
> > @lkraav
> > I fixed the search for Firefox. There's an issue when using `vaadin-text-field`, `@input`, and `debounce` where `event.target` would switch from the input to the custom element randomly. I used `event.path`, though did not test on Firefox. I switched to querying the element, and getting the value rather than relying on the event which I've tested in Chrome/Firefox (There's no Windows client for Safari) and it works.
> 
> Doesn't this work?
> 
> ```
> const path = e.composedPath  
>     ? e.composedPath()  
>     : (e.path || []);  
> ```

I created a reproduction [here](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcbmltcG9ydCB7Y3VzdG9tRWxlbWVudCwgcHJvcGVydHl9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcbmltcG9ydCB7IGRlYm91bmNlIH0gZnJvbSAnbG9kYXNoLWVzJztcblxuQGN1c3RvbUVsZW1lbnQoJ3NpbXBsZS1ncmVldGluZycpXG5leHBvcnQgY2xhc3MgU2ltcGxlR3JlZXRpbmcgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgc3RhdGljIHN0eWxlcyA9IGNzc2BgO1xuXG4gIEBwcm9wZXJ0eSgpXG4gIGN1cnJlbnRUYXJnZXQ7XG4gIFxuICBAcHJvcGVydHkoKVxuICBwYXRoO1xuICBcbiAgQHByb3BlcnR5KClcbiAgY29tcG9zZWRQYXRoO1xuXG4gIHJlbmRlcigpIHtcbiAgICByZXR1cm4gaHRtbGBcbiAgICAgICAgPHVsPlxuICAgICAgICAgIDxsaT5cbiAgICAgICAgICAgIFNpbXBsZTogPGlucHV0IEBpbnB1dD0ke3RoaXMub25JbnB1dH0gbGFiZWw9XCJTaW1wbGVcIiAvPlxuICAgICAgICAgIDwvbGk-XG4gICAgICAgICAgPGxpPlxuICAgICAgICAgICAgRGVib3VuY2VkOiA8aW5wdXQgQGlucHV0PSR7dGhpcy5kZWJvdW5jZWRPbklucHV0fSBsYWJlbD1cIkRlYm91bmNlZFwiIC8-XG4gICAgICAgICAgPC9saT5cbiAgICAgICAgPC91bD5cbiAgICAgICAgPGhyIC8-XG4gICAgICAgIDx1bD5cbiAgICAgICAgIDxsaT5cbiAgICAgICAgIEV2ZW50IFRhcmdldDogJHt0aGlzLmN1cnJlbnRUYXJnZXQ_LnRhZ05hbWV9XG4gICAgICAgIDwvbGk-XG4gICAgICAgIDxsaT5cbiAgICAgICAgIFBhdGhbMF06ICR7dGhpcy5wYXRoPy50YWdOYW1lfVxuICAgICAgICA8L2xpPlxuICAgICAgICA8bGk-XG4gICAgICAgICBDb21wb3NlZCBQYXRoOiAke3RoaXMuY29tcG9zZWRQYXRoPy50YWdOYW1lfVxuICAgICAgICA8L2xpPlxuICAgICAgICA8L2Rpdj5cbiAgICBgO1xuICB9XG4gIFxuICBjb25zdHJ1Y3RvcigpIHtcbiAgICBzdXBlcigpO1xuICAgIHRoaXMuZGVib3VuY2VkT25JbnB1dCA9IGRlYm91bmNlKHRoaXMub25JbnB1dCwgMzAwKTtcbiAgfVxuICBcbiAgb25JbnB1dChldmVudCkge1xuICAgICBkZWxldGUodGhpcy5jb21wb3NlZFBhdGgpO1xuICAgICBkZWxldGUodGhpcy5jdXJyZW50VGFyZ2V0KTtcbiAgICAgZGVsZXRlKHRoaXMucGF0aCk7XG4gICAgXG4gICAgIHRoaXMuY29tcG9zZWRQYXRoID0gZXZlbnQuY29tcG9zZWRQYXRoKClbMF07XG4gICAgIHRoaXMuY3VycmVudFRhcmdldCA9IGV2ZW50LnRhcmdldDtcbiAgICAgdGhpcy5wYXRoID0gZXZlbnQucGF0aFswXTtcbiAgfVxufVxuIn0seyJuYW1lIjoiaW5kZXguaHRtbCIsImNvbnRlbnQiOiI8IURPQ1RZUEUgaHRtbD5cbjxoZWFkPlxuICA8c2NyaXB0IHR5cGU9XCJtb2R1bGVcIiBzcmM9XCIuL3NpbXBsZS1ncmVldGluZy5qc1wiPjwvc2NyaXB0PlxuPC9oZWFkPlxuPGJvZHk-XG4gIDxzaW1wbGUtZ3JlZXRpbmcgbmFtZT1cIldvcmxkXCI-PC9zaW1wbGUtZ3JlZXRpbmc-XG48L2JvZHk-XG4ifSx7Im5hbWUiOiJwYWNrYWdlLmpzb24iLCJjb250ZW50Ijoie1xuICBcImRlcGVuZGVuY2llc1wiOiB7XG4gICAgXCJsaXRcIjogXCJeMi4wLjBcIixcbiAgICBcIkBsaXQvcmVhY3RpdmUtZWxlbWVudFwiOiBcIl4xLjAuMFwiLFxuICAgIFwibGl0LWVsZW1lbnRcIjogXCJeMy4wLjBcIixcbiAgICBcImxpdC1odG1sXCI6IFwiXjIuMC4wXCJcbiAgfVxufSIsImhpZGRlbiI6dHJ1ZX1d).

We can't rely on `composedPath`, `path`, or `target` when using an async method (https://lit-and-friends.slack.com/archives/C03PF4L4L/p1658504179662949).

saas786 commented:
> > > @lkraav
> > > I fixed the search for Firefox. There's an issue when using `vaadin-text-field`, `@input`, and `debounce` where `event.target` would switch from the input to the custom element randomly. I used `event.path`, though did not test on Firefox. I switched to querying the element, and getting the value rather than relying on the event which I've tested in Chrome/Firefox (There's no Windows client for Safari) and it works.
> > 
> > 
> > Doesn't this work?
> > ```
> > const path = e.composedPath  
> >     ? e.composedPath()  
> >     : (e.path || []);  
> > ```
> 
> @saas786
> 
> I created a reproduction [here](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQge2h0bWwsIGNzcywgTGl0RWxlbWVudH0gZnJvbSAnbGl0JztcbmltcG9ydCB7Y3VzdG9tRWxlbWVudCwgcHJvcGVydHl9IGZyb20gJ2xpdC9kZWNvcmF0b3JzLmpzJztcbmltcG9ydCB7IGRlYm91bmNlIH0gZnJvbSAnbG9kYXNoLWVzJztcblxuQGN1c3RvbUVsZW1lbnQoJ3NpbXBsZS1ncmVldGluZycpXG5leHBvcnQgY2xhc3MgU2ltcGxlR3JlZXRpbmcgZXh0ZW5kcyBMaXRFbGVtZW50IHtcbiAgc3RhdGljIHN0eWxlcyA9IGNzc2BgO1xuXG4gIEBwcm9wZXJ0eSgpXG4gIGN1cnJlbnRUYXJnZXQ7XG4gIFxuICBAcHJvcGVydHkoKVxuICBwYXRoO1xuICBcbiAgQHByb3BlcnR5KClcbiAgY29tcG9zZWRQYXRoO1xuXG4gIHJlbmRlcigpIHtcbiAgICByZXR1cm4gaHRtbGBcbiAgICAgICAgPHVsPlxuICAgICAgICAgIDxsaT5cbiAgICAgICAgICAgIFNpbXBsZTogPGlucHV0IEBpbnB1dD0ke3RoaXMub25JbnB1dH0gbGFiZWw9XCJTaW1wbGVcIiAvPlxuICAgICAgICAgIDwvbGk-XG4gICAgICAgICAgPGxpPlxuICAgICAgICAgICAgRGVib3VuY2VkOiA8aW5wdXQgQGlucHV0PSR7dGhpcy5kZWJvdW5jZWRPbklucHV0fSBsYWJlbD1cIkRlYm91bmNlZFwiIC8-XG4gICAgICAgICAgPC9saT5cbiAgICAgICAgPC91bD5cbiAgICAgICAgPGhyIC8-XG4gICAgICAgIDx1bD5cbiAgICAgICAgIDxsaT5cbiAgICAgICAgIEV2ZW50IFRhcmdldDogJHt0aGlzLmN1cnJlbnRUYXJnZXQ_LnRhZ05hbWV9XG4gICAgICAgIDwvbGk-XG4gICAgICAgIDxsaT5cbiAgICAgICAgIFBhdGhbMF06ICR7dGhpcy5wYXRoPy50YWdOYW1lfVxuICAgICAgICA8L2xpPlxuICAgICAgICA8bGk-XG4gICAgICAgICBDb21wb3NlZCBQYXRoOiAke3RoaXMuY29tcG9zZWRQYXRoPy50YWdOYW1lfVxuICAgICAgICA8L2xpPlxuICAgICAgICA8L2Rpdj5cbiAgICBgO1xuICB9XG4gIFxuICBjb25zdHJ1Y3RvcigpIHtcbiAgICBzdXBlcigpO1xuICAgIHRoaXMuZGVib3VuY2VkT25JbnB1dCA9IGRlYm91bmNlKHRoaXMub25JbnB1dCwgMzAwKTtcbiAgfVxuICBcbiAgb25JbnB1dChldmVudCkge1xuICAgICBkZWxldGUodGhpcy5jb21wb3NlZFBhdGgpO1xuICAgICBkZWxldGUodGhpcy5jdXJyZW50VGFyZ2V0KTtcbiAgICAgZGVsZXRlKHRoaXMucGF0aCk7XG4gICAgXG4gICAgIHRoaXMuY29tcG9zZWRQYXRoID0gZXZlbnQuY29tcG9zZWRQYXRoKClbMF07XG4gICAgIHRoaXMuY3VycmVudFRhcmdldCA9IGV2ZW50LnRhcmdldDtcbiAgICAgdGhpcy5wYXRoID0gZXZlbnQucGF0aFswXTtcbiAgfVxufVxuIn0seyJuYW1lIjoiaW5kZXguaHRtbCIsImNvbnRlbnQiOiI8IURPQ1RZUEUgaHRtbD5cbjxoZWFkPlxuICA8c2NyaXB0IHR5cGU9XCJtb2R1bGVcIiBzcmM9XCIuL3NpbXBsZS1ncmVldGluZy5qc1wiPjwvc2NyaXB0PlxuPC9oZWFkPlxuPGJvZHk-XG4gIDxzaW1wbGUtZ3JlZXRpbmcgbmFtZT1cIldvcmxkXCI-PC9zaW1wbGUtZ3JlZXRpbmc-XG48L2JvZHk-XG4ifSx7Im5hbWUiOiJwYWNrYWdlLmpzb24iLCJjb250ZW50Ijoie1xuICBcImRlcGVuZGVuY2llc1wiOiB7XG4gICAgXCJsaXRcIjogXCJeMi4wLjBcIixcbiAgICBcIkBsaXQvcmVhY3RpdmUtZWxlbWVudFwiOiBcIl4xLjAuMFwiLFxuICAgIFwibGl0LWVsZW1lbnRcIjogXCJeMy4wLjBcIixcbiAgICBcImxpdC1odG1sXCI6IFwiXjIuMC4wXCJcbiAgfVxufSIsImhpZGRlbiI6dHJ1ZX1d).
> 
> We can't rely on `composedPath`, `path`, or `target` when using an async method ([lit-and-friends.slack.com/archives/C03PF4L4L/p1658504179662949](https://lit-and-friends.slack.com/archives/C03PF4L4L/p1658504179662949)).

Wouldn't `async` logic work here? As we had same issue, and making it async resolved it.

anoblet commented:
"Some properties get lost if you access to them out of sync, unfortunately. That includes target and similar properties. You'll have to store them somehow."

lkraav commented:
> > I'm not sure I can visualize how a modal would work, could you please explain?
> 
> vaadin-dialog in the middle of screen. Not sure there's anything else to imagine. But on phone rn, so can't really browse vaadin element demos.

1. Looking again, I was off with the "middle of screen" idea - I didn't realize it would not auto-close after shifting video position.

Still, I believe https://vaadin.com/docs/latest/components/dialog could also be positioned where custom div sits now, with `modeless` attribute, so it continues to allow interaction with other elements.

In case vaadin-dialog with its built-in optimized design is out of reach for this iteration, "Chapters" title should be at least *bolded* and UPPERCASED to differentiate from chapter list..

![image](https://user-images.githubusercontent.com/147228/180985885-8592867a-cf1a-421a-808c-dc733599674c.png)


2. Should `.captions span` be `cursor: default`?

anoblet commented:
1) There's a fair amount of complexity with using `vaadin-dialog`. The actual dialog (`vaadin-dialog-overlay`) is recreated and appended to the document root each time it's opened.

I updated the chapters title to be uppercased and bold as well as added some padding and a faint bottom border.

![image](https://user-images.githubusercontent.com/7674171/181000905-a799c8dc-69dc-4699-b588-f09d50802356.png)

2) I wasn't sure whether or not to use `cursor: pointer` or `cursor: default`. @saas786 suggested we go with `default` (https://github.com/conversionxl/aybolit/pull/210#discussion_r903608885)

lkraav commented:
> I wasn't sure whether or not to use cursor: pointer or cursor: default

It is difficult to tell whether people use it more to select text or navigate video, but I would think that using the search box to find a keyword, then clicking on it would be one of the more involved use cases, where default cursor (text selection) solution looks weird.

saas786 commented:
> 2\. cursor: default

We can just remove it.

anoblet commented:
It's not best practice, though we could keep `cursor: default`, and use `cursor: pointer` for matched search terms?

saas786 commented:
> It's not best practice, though we could keep `cursor: default`, and use `cursor: pointer` for matched search terms?
> 
> There's also `cursor: alias` ([w3schools.com/cssref/playit.asp?filename=playcss_cursor&preval=alias](https://www.w3schools.com/cssref/playit.asp?filename=playcss_cursor&preval=alias))

`cursor: default` removed or having it shouldn't affect the UX.

We are already offering highlight, so that should be enough, no additional attention is needed I think.

anoblet commented:
@lkraav @kylebrodeur and I agree that `cursor: pointer` is more natural. That being said, I've created a CSS property allowing for it to be overridden.

```
jw-player {
  --captions-cursor: initial
}
```

anoblet commented:
Can we merge this in?
